packages/xo-server-load-balancer/src/plan.js

```diff
@@ -777,11 +809,311 @@ export default class Plan {
           bestVariance = variance
           bestVm = vm
         } else {
-          debugAffinity(`VM (${vm.id}) of Host (${sourceHost.id}) does not support pool migration.`)
+          debugAntiAffinity(`VM (${vm.id}) of Host (${sourceHost.id}) does not support pool migration.`)
         }
       }
     }
 
     return bestVm
   }
+
+  // ===================================================================
+  // Affinity helpers
+  // ===================================================================
+
+  async _processAffinity() {
+    if (!this._affinityTags.length) {
+      return
+    }
+
+    const allHosts = this._getHosts()
+    if (allHosts.length <= 1) {
+      return
+    }
+    const idToHost = keyBy(allHosts, 'id')
+
+    const allVms = filter(this._getAllRunningVms(), vm => vm.$container in idToHost)
+    const taggedHosts = this._getTaggedHosts({
+      hosts: allHosts,
+      tagList: this._affinityTags,
+      vms: allVms,
+      includeUntaggedVms: true,
+    })
+
+    // 1. Check if we must migrate VMs...
+    const spreadTags = []
+    for (const watchedTag of this._affinityTags) {
+      const taggedHostCount = taggedHosts.hosts.reduce(
+        (accumulator, host) => accumulator + (host.tags[watchedTag] > 0),
+        0
+      )
+      if (taggedHostCount > 1) {
+        spreadTags.push(watchedTag)
+      }
+    }
+    if (spreadTags.length === 0) {
+      return
+    }
+
+    // 2. Check for tag coalitions: when a VM has multiple affinity tags, these tags should be considered as the same tag
+    const coalitions = this._computeCoalitions(allVms)
+    const coalitionExample = Object.values(coalitions).find(coalition => coalition.length > 1)
+    if (coalitionExample !== undefined) {
+      warn(`affinity: Some VMs have multiple affinity tags, this should be avoided: ${coalitionExample}`)
```

Comment:
`coalitionExample` and `coalitions` are objects/arrays, will display as `[object Object]` , i suggest : 


```suggestion
      warn(`affinity: Some VMs have multiple affinity tags, this should be avoided: ${inspect(coalitionExample)}`)
      debugAffinity(`Tag coalitions: ${inspect(coalitions, { depth: null })}`)
```